include/bout/macro_for_each.hxx

```diff
@@ -0,0 +1,84 @@
+
+#ifndef __MACRO_FOR_EACH_H__
+#define __MACRO_FOR_EACH_H__
+
+// Provides a macro MACRO_FOR_EACH which applies a
+// macro to each argument in a VA_ARGS list
+// 
+// Based on this blog post by Daniel Hardman:
+//    https://codecraft.co/2014/11/25/variadic-macros-tricks/
+// This answer on StackOverflow:
+//    https://stackoverflow.com/questions/11761703/overloading-macro-on-number-of-arguments/11763277
+// See also: 
+//    https://github.com/cormacc/va_args_iterators
+//
+
+
+/// _me_x set of macros expand a number of arguments without ';' between them
+#define _me_1(_call, x) _call(x)
+#define _me_2(_call, x, ...) _call(x) _me_1(_call, __VA_ARGS__)
+#define _me_3(_call, x, ...) _call(x) _me_2(_call, __VA_ARGS__)
+#define _me_4(_call, x, ...) _call(x) _me_3(_call, __VA_ARGS__)
+#define _me_5(_call, x, ...) _call(x) _me_4(_call, __VA_ARGS__)
+#define _me_6(_call, x, ...) _call(x) _me_5(_call, __VA_ARGS__)
+#define _me_7(_call, x, ...) _call(x) _me_6(_call, __VA_ARGS__)
+#define _me_8(_call, x, ...) _call(x) _me_7(_call, __VA_ARGS__)
+#define _me_9(_call, x, ...) _call(x) _me_8(_call, __VA_ARGS__)
+#define _me_10(_call, x, ...) _call(x) _me_9(_call, __VA_ARGS__)
+
+/// _fe_x set of macros expand a number of arguments with ';' between them
+#define _fe_1(_call, x) _call(x);
+#define _fe_2(_call, x, ...) _call(x); _fe_1(_call, __VA_ARGS__)
+#define _fe_3(_call, x, ...) _call(x); _fe_2(_call, __VA_ARGS__)
+#define _fe_4(_call, x, ...) _call(x); _fe_3(_call, __VA_ARGS__)
+#define _fe_5(_call, x, ...) _call(x); _fe_4(_call, __VA_ARGS__)
+#define _fe_6(_call, x, ...) _call(x); _fe_5(_call, __VA_ARGS__)
+#define _fe_7(_call, x, ...) _call(x); _fe_6(_call, __VA_ARGS__)
+#define _fe_8(_call, x, ...) _call(x); _fe_7(_call, __VA_ARGS__)
+#define _fe_9(_call, x, ...) _call(x); _fe_8(_call, __VA_ARGS__)
+#define _fe_10(_call, x, ...) _call(x); _fe_9(_call, __VA_ARGS__)
+
+/// When called with __VA_ARGS__ first, this evaluates to an argument which depends
+/// on the length of __VA_ARGS__. This is used to find the appropriate macro to
+/// begin the expansion.
+#define _GET_FOR_EACH_EXPANSION(_1, _2, _3, _4, _5, _6, _7, _8, _9, _10, NAME, ...) NAME
+
+/// Apply a macro (first argument) to each
+/// of the following arguments.
+/// Currently supports up to 10 arguments.
+///
+/// Example:
+///
+///    MACRO_FOR_EACH(test, a, b, c)
+///
+/// expands to
+///
+///    test(a) test(b) test(c)
```

Comment:
I guess this form is most useful when `test` is actually another macro that itself includes `;`? Are there any other use cases?

Reply:
The main use for this (and the reason I went down this rabbit hole) was to declare lists of variables. In SD1D the reactions are integrated over the cells using Simpson's rule, so each rate is calculated three times, for the left, centre and right side of each cell:
https://github.com/boutproject/SD1D/blob/master/sd1d.cxx#L971
I'd like to add many more reactions, and don't want to have to repeat each one three times. I've improved this a bit:
https://github.com/boutproject/SD1D/blob/new-multifluid/reaction_elastic.cxx#L57
but there's still a lot of boilerplate code which could (perhaps) be removed. What I'd like is to be able to write:

```
INTEGRATE(i, weight, Ni, Nn, Ti, Tn, Vi, Vn) {
    BoutReal R = a0 * Ni * Nn * Cs0 * sqrt((16. / PI) * Ti) * Nnorm / Omega_ci;
    Fel[i] += weight * (Vi - Vn) * R;
    Eel[i] += (3. / 2) * weight * (Ti - Tn) * R;
}
```
which would expand to something like:
```
for (auto i : region(RGN_NOBNDRY)) {   /// << Maybe have region as argument
    QuadValue Ti(Ti, i), Ni(Ni, i), Vi(Vi, i), ... /// << Here MACRO_FOR_EACH
    for (int j : QR.indices) {   /// << Some kind of gensym for index j
        weight = QR[j];
        BoutReal Ti = Ti[j], Ni = Ni[j], ... /// << MACRO_FOR_EACH
        // User code
    }
}
```
so inside the expression the Field2D or Field3D variables in the arguments to `INTEGRATE` are declared as `BoutReal`s, shadowing original variables. I'm not sure if this would be great or terrible. Perhaps both.

Reply:
For the kind of (mis-)feature this enables, see the updated SD1D reaction rate integration code: https://github.com/boutproject/SD1D/blob/new-multifluid/reaction_elastic.cxx#L47

I will admit that the language abuse needed is quite appalling:
https://github.com/boutproject/SD1D/blob/new-multifluid/reaction.hxx#L117

This uses a trick from https://stackoverflow.com/questions/866012/is-there-a-way-to-define-variables-of-two-different-types-in-a-for-loop-initiali in which `for` loops are used to define variables in a new scope. This enables two nested loops with a bunch of variables introduced in both to be wrapped up into what looks like a single loop, but is actually 19 nested loops and two `if...else` clauses.